### PR TITLE
datetime: allow boundary values for interval.new

### LINKED ISSUE #8878: Inconsistent interval boudaries check
`datetime.interval` allowed range should include boundaries, yet it doesn't.

```lua
tarantool> datetime.interval.new{year=11759221}
---
- error: 'builtin/box/console.lua:415: value 11759221 of year is out of allowed range
    [-11759221, 11759221]'
...
```

Square brackets imply boundaries inclusion, yet the boundaries are excluded.

https://github.com/tarantool/tarantool/blob/ff57f990f359f6d7866c1947174d8ba0e97b1ea6/src/lua/datetime.lua#L297-L301

At the same time, datetime range check

https://github.com/tarantool/tarantool/blob/ff57f990f359f6d7866c1947174d8ba0e97b1ea6/src/lua/datetime.lua#L263

and C interval range check

https://github.com/tarantool/tarantool/blob/ff57f990f359f6d7866c1947174d8ba0e97b1ea6/src/lib/core/datetime.c#L695

include boundaries, as expected.

### PR DESCRIPTION
**Pull request by DifferentialOrange**
Before this patch, one couldn't create new datetime interval with boundary value from Lua. At the same time, it was possible to create such interval from Lua through addition and subtraction. C range verification allow to create boundary value intervals, error message also implies that they should be allowed. (See #8878 for more info.)

Closes #8878

NO_DOC=small bug fix

### CODE REVIEW

**Review by andreyaksenov — APPROVED** [review]

File: changelogs/unreleased/gh-8878-interval-boundaries-fix.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+## bugfix/datetime
+
+* Fixed a bug (raising false positive error) when creating new intervals with
+  range boundary values (gh-8878).
```
[the remark this answers is not in the document]
Comment (DifferentialOrange) [reply changelogs/unreleased/gh-8878-interval-boundaries-fix.md]: Fixed, thanks

File: test/app-tap/datetime.test.lua [hunk 2]
```diff
@@ -1316,6 +1319,29 @@ test:test("Time intervals creation - range checks", function(test)
...
+        {{ min = -MAX_MIN_RANGE }, '-6184879877160 minutes'},
+        {{ min =  MAX_MIN_RANGE }, '+6184879877160 minutes'},
+        {{ sec = -MAX_SEC_RANGE }, '-371092792629600 seconds'},
+        {{ sec =  MAX_SEC_RANGE }, '+371092792629600 seconds'},
+        {{ nsec = -MAX_NSEC_RANGE }, '+0 seconds, -2147483647 nanoseconds'},
+        {{ nsec =  MAX_NSEC_RANGE }, '+0 seconds, 2147483647 nanoseconds'},
+    }
+    for _, row in pairs(boundary_values) do
+        local args, str = unpack(row)
+        test:is(tostring(inew(args)), str)
```
Comment (Lord-KA) [inline test/app-tap/datetime.test.lua:1343]: Nit:
`%s/str/expected`
`%s/args/value`
The current phrasing is a bit less intuitive, but feel free to ignore.

Comment (Lord-KA) [reply test/app-tap/datetime.test.lua:1343]: To check that this is actually a boundary value, we can expect error for `+-1` values (with `assert_raises()`). I think it would be nice.

**Review by Lord-KA — APPROVED** [review]
@DifferentialOrange, thanks for the patch!
Overall, I don't see any major issues, so LGTM, but see two comments below.

File: test/app-tap/datetime.test.lua (hunk 2 above)
Comment (DifferentialOrange) [reply test/app-tap/datetime.test.lua]: Thanks, fixed both

File: test/app-tap/datetime.test.lua [hunk 3]
```diff
@@ -1316,6 +1319,54 @@ test:test("Time intervals creation - range checks", function(test)
...
+        {{ day = -math.floor(MAX_DAY_RANGE) }, '-4295055470 days'},
+        {{ day =  math.floor(MAX_DAY_RANGE) }, '+4295055470 days'},
+        {{ hour = -MAX_HOUR_RANGE }, '-103081331286 hours'},
+        {{ hour =  MAX_HOUR_RANGE }, '+103081331286 hours'},
+        {{ min = -MAX_MIN_RANGE }, '-6184879877160 minutes'},
+        {{ min =  MAX_MIN_RANGE }, '+6184879877160 minutes'},
+        {{ sec = -MAX_SEC_RANGE }, '-371092792629600 seconds'},
+        {{ sec =  MAX_SEC_RANGE }, '+371092792629600 seconds'},
+        {{ nsec = -MAX_NSEC_RANGE }, '+0 seconds, -2147483647 nanoseconds'},
+        {{ nsec =  MAX_NSEC_RANGE }, '+0 seconds, 2147483647 nanoseconds'},
```
Comment (ligurio) [inline test/app-tap/datetime.test.lua:1339]: Boundary values in messages are calculated manually and hardcoded now. I would build these message in runtime:
```
'-11759221 years' -> ('-%d'):format(MAX_YEAR_RANGE)
'+11759221 years' -> ('%d'):format(MAX_YEAR_RANGE) 
...
```
In latter case these messages will depend on real boundary values and automatically will be updated when we will update boundary values.

**Review by ligurio — APPROVED** [review]
I've no objections, so LGTM.
However, I left a minor comment, please take a look.

File: test/app-tap/datetime.test.lua (hunk 3 above)
Comment (Lord-KA) [reply test/app-tap/datetime.test.lua:1339]: Actually, same goes for +-1 test, it seems quite easy to automate and make it more compact as well.

Comment (DifferentialOrange) [reply test/app-tap/datetime.test.lua]: Reworked tests to be the most compact and automized

File: test/app-tap/datetime.test.lua [hunk 4]
```diff
@@ -1316,6 +1351,27 @@ test:test("Time intervals creation - range checks", function(test)
...
+
+    local range_boundary = {
+        year = MAX_YEAR_RANGE,
+        month = MAX_MONTH_RANGE,
+        week = MAX_WEEK_RANGE,
+        day =  MAX_DAY_RANGE,
+        hour = MAX_HOUR_RANGE,
+        min = MAX_MIN_RANGE,
+        sec = MAX_SEC_RANGE,
+        nsec = MAX_NSEC_RANGE,
```
Comment (ligurio) [inline test/app-tap/datetime.test.lua:1363]: Missed `MAX_USEC_RANGE` and `MAX_MSEC_RANGE`.
Both could be used in intervals:

```
tarantool>  datetime.interval.new{msec=2147}
---
- error: 'builtin/box/console.lua:415: value 2147 of msec is out of allowed range
    [-2147, 2147]'
...

tarantool>  datetime.interval.new{usec=2147}
---
- +0 seconds, 2147000 nanoseconds
...

tarantool> 
```

File: test/app-tap/datetime.test.lua (hunk 3 above)
Comment (ligurio) [reply test/app-tap/datetime.test.lua]: Thanks!

File: test/app-tap/datetime.test.lua (hunk 4 above)
Comment (DifferentialOrange) [reply test/app-tap/datetime.test.lua:1373]: Added

**Comment by igormunkin** [comment]
Also backported to 2.11 and 2.10 in scope of 230aba5 and 3af2b56 respectively.
